// @uniformdev/canvas-react/dist/index.mjs
// src/components/DefaultNotImplementedComponent.tsx
import { CANVAS_LOCALIZATION_TYPE } from "@uniformdev/canvas";
import React from "react";
var wrapperStyles = {
  display: "flex",
  flexDirection: "column",
  gap: "0.5rem",
  borderLeft: "4px solid #e42535",
  padding: "16px",
  fontSize: "16px",
  borderRadius: "0 8px 8px 0",
  margin: "8px",
  backgroundColor: "rgba(255, 255, 255, 0.45)",
  color: "#1d3557"
};
function DefaultNotImplementedComponent(props) {
  var _a, _b, _c;
  const componentType = (_a = props.component) == null ? void 0 : _a.type;
  if (!componentType) {
    return null;
  }
  if (componentType === CANVAS_LOCALIZATION_TYPE) {
    return /* @__PURE__ */ React.createElement("div", { style: wrapperStyles }, /* @__PURE__ */ React.createElement("p", null, "Seems like localization is not enabled in your application. Please read our documentation on how to", " ", /* @__PURE__ */ React.createElement(
      "a",
      {
        href: "https://docs.uniform.app/guides/composition/localization#activate-front-end",
        target: "_blank",
        style: { fontWeight: "bolder", textDecoration: "underline" },
        rel: "noreferrer"
      },
      "enable localization in your front-end application."
    )));
  }
  const proposedFileName = `${componentType[0].toUpperCase()}${componentType.substring(1)}`;
  const probableProps = Object.keys((_b = props.component.parameters) != null ? _b : {});
  const probableSlots = Object.keys((_c = props.component.slots) != null ? _c : {});
  return /* @__PURE__ */ React.createElement("div", { style: wrapperStyles }, /* @__PURE__ */ React.createElement("h2", { style: { fontSize: "1.2rem" } }, "Unknown Component"), /* @__PURE__ */ React.createElement("p", null, "Received request from Uniform to render a component with the public ID: ", /* @__PURE__ */ React.createElement("code", null, componentType), "."), /* @__PURE__ */ React.createElement("p", null, /* @__PURE__ */ React.createElement("code", null, "<UniformComposition />"), " does not have ", /* @__PURE__ */ React.createElement("code", null, componentType), " mapped to a React component yet."), /* @__PURE__ */ React.createElement("p", null, "To teach your app how to render this component:"), " ", /* @__PURE__ */ React.createElement(
    "ul",
    {
      style: {
        listStyleType: "disc",
        marginLeft: "1rem",
        display: "flex",
        flexDirection: "column",
        gap: "0.5rem"
      }
    },
    /* @__PURE__ */ React.createElement("li", null, "Create a React component and register it with Uniform, for example", /* @__PURE__ */ React.createElement("pre", null, `function ${proposedFileName}(${probableProps.length > 0 ? `{ ${probableProps.join(", ")} }` : ""}) {
  return (
    <div>
      ${proposedFileName}!
      ${probableSlots.length > 0 ? probableSlots.map((slot) => `<UniformSlot name="${slot}" />`).join("\n      ") : ""}
    </div>
  )
}

registerUniformComponent({ type: "${componentType}", component: ${proposedFileName} })`), /* @__PURE__ */ React.createElement("details", null, /* @__PURE__ */ React.createElement("summary", { style: { cursor: "pointer" } }, "Props that your React component will receive"), /* @__PURE__ */ React.createElement("pre", null, JSON.stringify(props, null, 2)))),
    /* @__PURE__ */ React.createElement("li", null, "Import the component into the file where ", /* @__PURE__ */ React.createElement("code", null, "<UniformComposition />"), " is defined, for example ", /* @__PURE__ */ React.createElement("pre", null, `import "../components/${proposedFileName}.tsx"`))
  ), " ", /* @__PURE__ */ React.createElement("p", null, "Need more help?", " ", /* @__PURE__ */ React.createElement(
    "a",
    {
      href: "https://docs.uniform.app/docs/guides/composition/rendering",
      target: "_blank",
      rel: "noreferrer",
      style: { textDecoration: "underline" }
    },
    "Check out the documentation."
  )));
}

// src/components/UniformComponent.tsx
import {
  CANVAS_ENRICHMENT_TAG_PARAM
} from "@uniformdev/canvas";
import { Track, TrackFragment, useUniformContext } from "@uniformdev/context-react";
import React6, { createContext as createContext2, useContext as useContext2 } from "react";

// src/convertComponentToProps.ts
function convertComponentToProps(component) {
  var _a;
  const parameters = (_a = component.parameters) != null ? _a : {};
  const renderComponentProps = {
    ...Object.keys(parameters).reduce((acc, cur) => {
      acc[cur] = parameters[cur].value;
      return acc;
    }, {}),
    ...component.data,
    component
  };
  return renderComponentProps;
}

// src/storeDefinition.ts
var NOT_IMPLEMENTED_COMPONENT = "__not_implemented__";
var getTypeWithVariant = (type, variantId) => `${type}${variantId ? `__${variantId}` : ""}`;
var createComponentStore = () => {
  const components = /* @__PURE__ */ new Map();
  components.set(NOT_IMPLEMENTED_COMPONENT, DefaultNotImplementedComponent);
  return {
    register: ({ type, component, variantId }) => {
      components.set(getTypeWithVariant(type, variantId), component);
    },
    get: (type, variantId) => {
      return components.get(getTypeWithVariant(type, variantId));
    }
  };
};
var createComponentStoreResolver = ({ store, defaultComponent = DefaultNotImplementedComponent }) => {
  return (component) => {
    var _a;
    const resolved = (_a = store.get(getTypeWithVariant(component.type, component.variant))) != null ? _a : store.get(getTypeWithVariant(component.type));
    return resolved || defaultComponent;
  };
};

// src/store.ts
var componentStore = createComponentStore();
var registerUniformComponent = ({
  type,
  variantId,
  component
}) => {
  componentStore.register({
    type,
    variantId,
    component
  });
};
var componentStoreResolver = createComponentStoreResolver({
  store: componentStore
});

// src/components/UniformSlot.tsx
import {
  CANVAS_PERSONALIZE_SLOT,
  CANVAS_PERSONALIZE_TYPE,
  CANVAS_TEST_SLOT,
  CANVAS_TEST_TYPE,
  EMPTY_COMPOSITION as EMPTY_COMPOSITION2
} from "@uniformdev/canvas";
import React5 from "react";

// src/defaultSystemComponentResolver.tsx
import {
  mapSlotToPersonalizedVariations,
  mapSlotToTestVariations
} from "@uniformdev/canvas";
import { Personalize, Test } from "@uniformdev/context-react";
import * as React2 from "react";
var defaultSystemComponentResolver = {
  test: (component, key, renderChild) => {
    var _a, _b, _c, _d, _e;
    const testComponent = component;
    const variants = (_b = (_a = testComponent.slots) == null ? void 0 : _a.test) != null ? _b : [];
    const testName = (_e = (_d = (_c = testComponent.parameters) == null ? void 0 : _c.test) == null ? void 0 : _d.value) != null ? _e : "Untitled Test";
    const finalVariants = mapSlotToTestVariations(variants);
    return /* @__PURE__ */ React2.createElement(
      Test,
      {
        key,
        variations: finalVariants,
        name: testName,
        component: (variation) => renderChild(variation, key)
      }
    );
  },
  personalization: (component, key, renderChild) => {
    var _a, _b, _c, _d, _e, _f, _g;
    const pzComponent = component;
    const processedVariants = mapSlotToPersonalizedVariations((_a = pzComponent == null ? void 0 : pzComponent.slots) == null ? void 0 : _a.pz);
    return /* @__PURE__ */ React2.createElement(
      Personalize,
      {
        key,
        variations: processedVariants,
        count: Number((_d = (_c = (_b = pzComponent.parameters) == null ? void 0 : _b.count) == null ? void 0 : _c.value) != null ? _d : 1),
        name: (_g = (_f = (_e = pzComponent.parameters) == null ? void 0 : _e.trackingEventName) == null ? void 0 : _f.value) != null ? _g : "Untitled Personalization",
        component: (variation) => renderChild(variation, 0)
      }
    );
  }
};

// src/components/ContextualEditingComponentWrapper.tsx
import {
  CANVAS_LOCALE_TAG_PARAM,
  IN_CONTEXT_EDITOR_COMPONENT_END_ROLE,
  IN_CONTEXT_EDITOR_COMPONENT_START_ROLE,
  PLACEHOLDER_ID
} from "@uniformdev/canvas";
import React4 from "react";

// src/components/UniformComposition.tsx
import React3, { createContext, useContext } from "react";

// src/hooks/useUniformContextualEditing.ts
import {
  createCanvasChannel,
  createUniformApiEnhancer,
  EMPTY_COMPOSITION,
  IN_CONTEXT_EDITOR_EMBED_SCRIPT_ID,
  IN_CONTEXT_EDITOR_QUERY_STRING_PARAM,
  isUpdateCompositionInternalMessage
} from "@uniformdev/canvas";
import { useEffect, useMemo, useState } from "react";
var registeredCompositionIds = /* @__PURE__ */ new Set();
var useUniformContextualEditing = ({
  initialCompositionValue,
  enhance = (message) => message.composition
}) => {
  const [contextualComposition, setContextualComposition] = useState();
  const channel = useMemo(() => {
    var _a;
    if (!isInContextEditingMode()) {
      return;
    }
    const channel2 = createCanvasChannel({
      broadcastTo: [(_a = window.opener) != null ? _a : window.top],
      listenTo: [window]
    });
    return channel2;
  }, []);
  useEffect(() => {
    if ((contextualComposition == null ? void 0 : contextualComposition._id) && (initialCompositionValue == null ? void 0 : initialCompositionValue._id) !== EMPTY_COMPOSITION._id && (initialCompositionValue == null ? void 0 : initialCompositionValue._id) !== (contextualComposition == null ? void 0 : contextualComposition._id)) {
      setContextualComposition(void 0);
      return;
    }
    if (!channel || registeredCompositionIds.has(initialCompositionValue == null ? void 0 : initialCompositionValue._id)) {
      return;
    }
    const unsubscribeFromCompositionUpdates = channel.on("update-composition-internal", async (message) => {
      if (!isUpdateCompositionInternalMessage(message)) {
        return;
      }
      const enhancedComposition = await enhance(message);
      setContextualComposition(enhancedComposition);
    });
    registeredCompositionIds.add(initialCompositionValue == null ? void 0 : initialCompositionValue._id);
    return () => {
      unsubscribeFromCompositionUpdates();
      registeredCompositionIds.delete(initialCompositionValue == null ? void 0 : initialCompositionValue._id);
    };
  }, [channel, enhance, initialCompositionValue == null ? void 0 : initialCompositionValue._id, contextualComposition == null ? void 0 : contextualComposition._id]);
  useEffect(() => {
    if (!isInContextEditingMode()) {
      return;
    }
    const existingScript = document.getElementById(IN_CONTEXT_EDITOR_EMBED_SCRIPT_ID);
    if (existingScript) {
      return;
    }
    window.__UNIFORM_CONTEXTUAL_EDITING__ = {
      framework: "React",
      // Make sure to also update the value in canvas-vue
      version: 1
    };
    const script = document.createElement("script");
    script.id = IN_CONTEXT_EDITOR_EMBED_SCRIPT_ID;
    script.src = getCanvasInContextEmbedScriptUrl();
    script.async = true;
    document.head.appendChild(script);
  }, [initialCompositionValue == null ? void 0 : initialCompositionValue._id]);
  return {
    composition: contextualComposition != null ? contextualComposition : initialCompositionValue,
    isContextualEditing: Boolean(contextualComposition)
  };
};
function getCanvasInContextEmbedScriptUrl() {
  const scriptUrl = `${window.document.referrer}files/canvas-in-context-embed/index.js`;
  return scriptUrl;
}
function isInContextEditingMode() {
  if (typeof window === "undefined") {
    return false;
  }
  const isOpenedByInContextEditor = new URLSearchParams(window.location.search).has(
    IN_CONTEXT_EDITOR_QUERY_STRING_PARAM
  );
  const isAllowlistedReferrer = Boolean(
    window.document.referrer.match(/(^https:\/\/|\.)(uniform.app|uniform.wtf|localhost:\d{4})\//)
  );
  return isOpenedByInContextEditor && isAllowlistedReferrer;
}

// src/components/UniformComposition.tsx
var UniformCompositionContext = createContext({
  data: void 0,
  isContextualEditing: false
});
function useUniformCurrentComposition() {
  return useContext(UniformCompositionContext);
}
function UniformComposition({
  data,
  behaviorTracking = "onView",
  children,
  resolveRenderer,
  contextualEditingEnhancer
}) {
  const { composition, isContextualEditing } = useUniformContextualEditing({
    initialCompositionValue: data,
    enhance: contextualEditingEnhancer
  });
  return /* @__PURE__ */ React3.createElement(
    UniformCompositionContext.Provider,
    {
      value: { data: composition, behaviorTracking, resolveRenderer, isContextualEditing }
    },
    /* @__PURE__ */ React3.createElement(ContextualEditingComponentWrapper, { component: composition }, /* @__PURE__ */ React3.createElement(
      UniformComponent,
      {
        data: composition,
        behaviorTracking,
        resolveRenderer
      },
      children
    ))
  );
}

// src/components/ContextualEditingComponentWrapper.tsx
function ContextualEditingComponentWrapper({
  component,
  parentComponent,
  slotName,
  indexInSlot,
  slotChildrenCount,
  emptyPlaceholder,
  children
}) {
  var _a, _b, _c, _d;
  const isPlaceholder = (component == null ? void 0 : component._id) === PLACEHOLDER_ID;
  const { isContextualEditing } = useUniformCurrentComposition();
  return !isContextualEditing ? /* @__PURE__ */ React4.createElement(React4.Fragment, null, children) : isPlaceholder && emptyPlaceholder === null ? null : /* @__PURE__ */ React4.createElement(React4.Fragment, null, /* @__PURE__ */ React4.createElement(
    "script",
    {
      "data-role": IN_CONTEXT_EDITOR_COMPONENT_START_ROLE,
      "data-parent-id": parentComponent == null ? void 0 : parentComponent._id,
      "data-parent-type": parentComponent == null ? void 0 : parentComponent.type,
      "data-component-id": component == null ? void 0 : component._id,
      "data-slot-name": slotName != null ? slotName : "",
      "data-component-index": indexInSlot != null ? indexInSlot : "",
      "data-total-components": slotChildrenCount != null ? slotChildrenCount : "",
      "data-component-name": component == null ? void 0 : component.type,
      "data-is-placeholder": isPlaceholder ? "true" : void 0,
      "data-is-localized": ((_a = component == null ? void 0 : component.parameters) == null ? void 0 : _a[CANVAS_LOCALE_TAG_PARAM]) ? "true" : void 0,
      "data-component-title": (_d = (_c = (_b = component == null ? void 0 : component.parameters) == null ? void 0 : _b.title) == null ? void 0 : _c.value) != null ? _d : ""
    }
  ), isPlaceholder && emptyPlaceholder !== void 0 ? emptyPlaceholder : children, /* @__PURE__ */ React4.createElement("script", { "data-role": IN_CONTEXT_EDITOR_COMPONENT_END_ROLE }));
}

// src/components/UniformSlot.tsx
function UniformSlot({
  name,
  resolveRenderer,
  children,
  emptyPlaceholder
}) {
  var _a;
  const { data: parentData, resolveRenderer: parentResolveRenderer } = useUniformCurrentComponent();
  if (!parentData) {
    throw new Error("Cannot use Slot without a <UniformComponent /> wrapper.");
  }
  const slot = (_a = parentData.slots) == null ? void 0 : _a[name];
  if (!slot || !Array.isArray(slot)) {
    if (process.env.NODE_ENV === "development" && parentData.type !== EMPTY_COMPOSITION2.type) {
      console.warn(
        `[canvas-dev] slot '${name}' was rendered, but it was not defined on its component. This is expected if the slot is optional, otherwise it may indicate a typo. Component:`,
        parentData
      );
    }
    return null;
  }
  const resolver = resolveRenderer != null ? resolveRenderer : parentResolveRenderer;
  const systemResolver = defaultSystemComponentResolver;
  const finalChildren = slot.map((component, index) => {
    const child = renderComponent({
      component,
      resolveRenderer: resolver,
      resolveSystem: systemResolver,
      key: `inner-${index}`,
      indexInSlot: index,
      slotName: name,
      parentComponent: parentData,
      slotChildrenCount: slot.length,
      emptyPlaceholder
    });
    const elements = children ? children({ child, component, key: `wrapped-inner-${index}` }) : child;
    return React5.createElement(React5.Fragment, { key: index }, elements);
  });
  return React5.createElement(React5.Fragment, void 0, finalChildren);
}
function renderComponent({
  component,
  resolveRenderer,
  resolveSystem,
  key = 0,
  indexInSlot,
  slotName,
  parentComponent,
  slotChildrenCount,
  emptyPlaceholder
}) {
  const RenderComponent = resolveRenderer == null ? void 0 : resolveRenderer(component);
  if (component.type === CANVAS_TEST_TYPE) {
    const testComponent = resolveSystem.test(
      component,
      key,
      (variantComponent, key2) => {
        var _a, _b;
        return renderComponent({
          component: variantComponent,
          resolveRenderer,
          resolveSystem,
          key: key2,
          parentComponent: component,
          slotName: CANVAS_TEST_SLOT,
          slotChildrenCount: (_a = component == null ? void 0 : component.slots) == null ? void 0 : _a[CANVAS_TEST_SLOT].length,
          indexInSlot: (_b = component == null ? void 0 : component.slots) == null ? void 0 : _b[CANVAS_TEST_SLOT].findIndex(
            ({ _id }) => variantComponent._id === _id
          )
        });
      }
    );
    return /* @__PURE__ */ React5.createElement(
      ContextualEditingComponentWrapper,
      {
        component,
        parentComponent,
        slotName,
        indexInSlot,
        slotChildrenCount,
        emptyPlaceholder
      },
      testComponent
    );
  } else if (component.type === CANVAS_PERSONALIZE_TYPE) {
    const personalizationComponent = resolveSystem.personalization(
      component,
      key,
      (variantComponent, key2) => {
        var _a, _b;
        return renderComponent({
          component: variantComponent,
          resolveRenderer,
          resolveSystem,
          key: key2,
          parentComponent: component,
          slotName: CANVAS_PERSONALIZE_SLOT,
          slotChildrenCount: (_a = component == null ? void 0 : component.slots) == null ? void 0 : _a[CANVAS_PERSONALIZE_SLOT].length,
          indexInSlot: (_b = component == null ? void 0 : component.slots) == null ? void 0 : _b[CANVAS_PERSONALIZE_SLOT].findIndex(
            ({ _id }) => variantComponent._id === _id
          )
        });
      }
    );
    return /* @__PURE__ */ React5.createElement(
      ContextualEditingComponentWrapper,
      {
        component,
        parentComponent,
        slotName,
        indexInSlot,
        slotChildrenCount,
        emptyPlaceholder
      },
      personalizationComponent
    );
  } else if (RenderComponent) {
    const props = convertComponentToProps(component);
    return /* @__PURE__ */ React5.createElement(UniformComponent, { key, data: component, resolveRenderer }, /* @__PURE__ */ React5.createElement(
      ContextualEditingComponentWrapper,
      {
        component,
        parentComponent,
        slotName,
        indexInSlot,
        slotChildrenCount,
        emptyPlaceholder
      },
      /* @__PURE__ */ React5.createElement(RenderComponent, { ...props })
    ));
  } else if (process.env.NODE_ENV !== "production") {
    console.warn(
      `[canvas] found component of type '${component.type}'${component.variant ? ` with variant '${component.variant}'` : ""}. Nothing will be rendered. Check your resolveRenderer function or add registerUniformComponent to your component.
 See documentation https://docs.uniform.app/guides/dev/rendering`,
      component
    );
  }
  return null;
}

// src/components/UniformComponent.tsx
var UniformComponentContext = createContext2({});
function useUniformCurrentComponent() {
  return useContext2(UniformComponentContext);
}
function UniformComponent({
  data,
  resolveRenderer,
  children,
  behaviorTracking
}) {
  var _a, _b, _c;
  const parentData = useUniformCurrentComponent();
  const contextContextProviderPresent = useUniformContext({ throwOnMissingProvider: false }) !== void 0;
  if (!data) {
    if (process.env.NODE_ENV === "development") {
      console.warn(`[canvas-dev] UniformComponent was rendered with no data, nothing will be output.`);
    }
    return null;
  }
  const contextValue = {
    data,
    resolveRenderer: resolveRenderer || parentData.resolveRenderer || componentStoreResolver,
    behaviorTracking: (_a = behaviorTracking != null ? behaviorTracking : parentData == null ? void 0 : parentData.behaviorTracking) != null ? _a : "onView"
  };
  const enrichmentTags = (_c = (_b = data.parameters) == null ? void 0 : _b[CANVAS_ENRICHMENT_TAG_PARAM]) == null ? void 0 : _c.value;
  const TrackComponent = contextValue.behaviorTracking === "onLoad" ? TrackFragment : Track;
  const resolvedChildren = resolveChildren({
    children,
    data,
    hasParentLayout: Boolean(parentData.data),
    resolveRenderer: contextValue.resolveRenderer
  });
  return /* @__PURE__ */ React6.createElement(UniformComponentContext.Provider, { value: contextValue }, contextContextProviderPresent ? (
    /* auto-track behavior signals when in a Canvas context */
    /* @__PURE__ */ React6.createElement(TrackComponent, { behavior: enrichmentTags }, resolvedChildren)
  ) : resolvedChildren);
}
function resolveChildren({
  children,
  data,
  hasParentLayout,
  resolveRenderer
}) {
  var _a;
  if (!children && !hasParentLayout) {
    const topLevelComponent = resolveRenderer({ type: data.type });
    if (topLevelComponent) {
      children = React6.createElement(topLevelComponent, convertComponentToProps(data));
    } else {
      if (Object.keys((_a = data.slots) != null ? _a : {}).length > 1 && process.env.NODE_ENV === "development") {
        console.warn(
          `[canvas-dev] All the slots in component '${data.type}' are rendered in no particular order. Use '<UniformSlot name={slotName} />' to reliably render the slots.`
        );
      }
      children = Object.keys(data.slots || {}).map((slotName) => /* @__PURE__ */ React6.createElement(UniformSlot, { key: slotName, name: slotName }));
    }
  }
  const renderChildren = typeof children === "function" ? children(convertComponentToProps(data)) : children;
  return renderChildren;
}

// src/components/UniformText.tsx
import React7, { useCallback, useMemo as useMemo2, useState as useState2 } from "react";
var UniformText = ({
  as: Tag = "span",
  parameterId,
  isMultiline = false,
  placeholder,
  render = (value) => value,
  ...props
}) => {
  const { data: componentData } = useUniformCurrentComponent();
  const { isContextualEditing } = useUniformCurrentComposition();
  const [isFocused, setIsFocused] = useState2(false);
  const parameter = useMemo2(() => {
    var _a;
    return (_a = componentData == null ? void 0 : componentData.parameters) == null ? void 0 : _a[parameterId];
  }, [componentData, parameterId]);
  const value = useMemo2(() => parameter == null ? void 0 : parameter.value, [parameter]);
  const isEditable = useMemo2(() => {
    var _a, _b;
    return (_b = (_a = parameter == null ? void 0 : parameter._contextualEditing) == null ? void 0 : _a.isEditable) != null ? _b : false;
  }, [parameter]);
  const handleOnFocus = useCallback(() => {
    setIsFocused(true);
  }, [setIsFocused]);
  const handleOnBlur = useCallback(() => {
    setIsFocused(false);
  }, [setIsFocused]);
  const shouldSkipCustomRendering = useMemo2(() => isFocused && isEditable, [isFocused, isEditable]);
  if (!parameter) {
    console.warn(
      `UniformText: The parameterId "${parameterId}" was not found in the component of type "${componentData == null ? void 0 : componentData.type}"`
    );
    return null;
  }
  if (!isContextualEditing) {
    return /* @__PURE__ */ React7.createElement(Tag, { style: isMultiline ? { whiteSpace: "pre-wrap" } : {}, ...props }, render(value));
  }
  return /* @__PURE__ */ React7.createElement(
    Tag,
    {
      ...props,
      "data-uniform-component-id": componentData == null ? void 0 : componentData._id,
      "data-uniform-parameter-id": parameterId,
      "data-uniform-parameter-value": value != null ? value : "",
      "data-uniform-parameter-type": "text",
      "data-uniform-is-multiline": isMultiline,
      "data-uniform-placeholder": placeholder,
      style: isMultiline ? { whiteSpace: "pre-wrap" } : {},
      contentEditable: isEditable,
      suppressContentEditableWarning: true,
      onFocus: handleOnFocus,
      onBlur: handleOnBlur
    },
    shouldSkipCustomRendering ? value : render(value)
  );
};

// src/helpers/getParameterAttributes.ts
var getParameterAttributes = ({ id, component, placeholder }) => {
  var _a, _b, _c;
  const componentId = component == null ? void 0 : component._id;
  const parameter = (_a = component == null ? void 0 : component.parameters) == null ? void 0 : _a[id];
  const value = parameter == null ? void 0 : parameter.value;
  const type = parameter == null ? void 0 : parameter.type;
  const isEditable = (_c = (_b = parameter == null ? void 0 : parameter._contextualEditing) == null ? void 0 : _b.isEditable) != null ? _c : false;
  return {
    "data-uniform-component-id": componentId,
    "data-uniform-parameter-id": id,
    "data-uniform-parameter-value": value != null ? value : "",
    "data-uniform-parameter-type": type,
    "data-uniform-placeholder": placeholder,
    contentEditable: isEditable,
    suppressContentEditableWarning: true
  };
};

// src/hooks/useCompositionEventEffect.ts
import {
  CANVAS_DRAFT_STATE,
  createEventBus,
  subscribeToComposition
} from "@uniformdev/canvas";
import { useEffect as useEffect2 } from "react";
function useCompositionEventEffect({
  enabled,
  projectId,
  compositionId,
  effect
}) {
  useEffect2(() => {
    if (!enabled || !compositionId || !projectId) {
      return;
    }
    let goodbye = void 0;
    const loadEffect = async () => {
      const eventBus = await createEventBus();
      if (eventBus) {
        goodbye = subscribeToComposition({
          eventBus,
          compositionId,
          compositionState: CANVAS_DRAFT_STATE,
          projectId,
          callback: effect,
          event: "updated"
        });
      }
    };
    loadEffect();
    return () => {
      if (goodbye) {
        goodbye();
      }
    };
  }, [compositionId, enabled, projectId, effect]);
}
export {
  DefaultNotImplementedComponent,
  NOT_IMPLEMENTED_COMPONENT,
  UniformComponent,
  UniformComposition,
  UniformSlot,
  UniformText,
  componentStore,
  componentStoreResolver,
  createComponentStore,
  createComponentStoreResolver,
  createUniformApiEnhancer,
  getParameterAttributes,
  registerUniformComponent,
  useCompositionEventEffect,
  useUniformContextualEditing,
  useUniformCurrentComponent,
  useUniformCurrentComposition
};
