// @uniformdev/canvas-next/dist/slug/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);

// src/slug/index.ts
var slug_exports = {};
__export(slug_exports, {
  getServerSideProps: () => getServerSideProps,
  getStaticPaths: () => getStaticPaths,
  getStaticProps: () => getStaticProps,
  withUniformGetServerSideProps: () => withUniformGetServerSideProps,
  withUniformGetStaticPaths: () => withUniformGetStaticPaths,
  withUniformGetStaticProps: () => withUniformGetStaticProps
});
module.exports = __toCommonJS(slug_exports);

// src/slug/withUniformGetServerSideProps.ts
var import_canvas = require("@uniformdev/canvas");
var withUniformGetServerSideProps = (options) => {
  const canvasClient = (options == null ? void 0 : options.client) || new import_canvas.CanvasClient({
    apiKey: process.env.UNIFORM_API_KEY,
    projectId: process.env.UNIFORM_PROJECT_ID,
    apiHost: process.env.UNIFORM_CLI_BASE_URL,
    edgeApiHost: process.env.UNIFORM_CLI_BASE_EDGE_URL
  });
  return async function wrappedGetServerSideProps(context) {
    var _a;
    const { preview, previewData } = context;
    let composition = void 0;
    let slug = (options == null ? void 0 : options.prefix) ? context.resolvedUrl.replace(new RegExp(`^${options.prefix}`), "") : context.resolvedUrl;
    if (options == null ? void 0 : options.modifySlug) {
      slug = options.modifySlug(slug, context);
    }
    if (previewData == null ? void 0 : previewData.isUniformContextualEditing) {
      composition = { ...import_canvas.EMPTY_COMPOSITION, _id: (_a = previewData.compositionId) != null ? _a : import_canvas.EMPTY_COMPOSITION._id };
    } else {
      try {
        const time = Date.now();
        const response = await canvasClient.getCompositionBySlug({
          ...options == null ? void 0 : options.requestOptions,
          slug,
          state: preview || (options == null ? void 0 : options.preview) ? import_canvas.CANVAS_DRAFT_STATE : import_canvas.CANVAS_PUBLISHED_STATE
        });
        const duration = Date.now() - time;
        composition = response.composition;
        if (!(options == null ? void 0 : options.silent)) {
          (0, import_canvas.logCompositionResponse)(response, duration);
        }
      } catch (e) {
        console.error("[canvas-next] Failed to fetch composition", e);
        return {
          notFound: true
        };
      }
    }
    const ret = (options == null ? void 0 : options.callback) ? await options.callback(context, composition) : { props: {} };
    if (composition && Object.hasOwn(ret, "props")) {
      const casted = ret;
      casted.props["data"] = composition;
    }
    return ret;
  };
};

// src/slug/withUniformGetStaticPaths.ts
var import_canvas2 = require("@uniformdev/canvas");
var withUniformGetStaticPaths = (options) => {
  return async function wrappedGetStaticPaths() {
    var _a;
    const canvasClient = (_a = options == null ? void 0 : options.client) != null ? _a : new import_canvas2.CanvasClient({
      apiKey: process.env.UNIFORM_API_KEY,
      projectId: process.env.UNIFORM_PROJECT_ID,
      apiHost: process.env.UNIFORM_CLI_BASE_URL,
      edgeApiHost: process.env.UNIFORM_CLI_BASE_EDGE_URL
    });
    const response = await canvasClient.getCompositionList({
      ...options == null ? void 0 : options.requestOptions,
      state: (options == null ? void 0 : options.preview) ? import_canvas2.CANVAS_DRAFT_STATE : import_canvas2.CANVAS_PUBLISHED_STATE
    });
    const compositions = (options == null ? void 0 : options.callback) ? await options.callback(response.compositions) : response.compositions;
    const paths = compositions.filter((reference) => !!reference.composition._slug).map((composition) => {
      var _a2;
      return `${(_a2 = options == null ? void 0 : options.prefix) != null ? _a2 : ""}${composition.composition._slug}`;
    });
    return {
      paths,
      fallback: true
    };
  };
};

// src/slug/withUniformGetStaticProps.ts
var import_canvas3 = require("@uniformdev/canvas");

// src/helpers/resolveSlugFromParams.ts
var resolveSlugFromParams = ({
  param = "slug",
  params
}) => {
  const slug = (params == null ? void 0 : params[param]) || "";
  const slugString = Array.isArray(slug) ? slug.join("/") : slug;
  return `/${slugString}`;
};

// src/slug/withUniformGetStaticProps.ts
var withUniformGetStaticProps = (options) => {
  var _a;
  const canvasClient = (_a = options == null ? void 0 : options.client) != null ? _a : new import_canvas3.CanvasClient({
    apiKey: process.env.UNIFORM_API_KEY,
    projectId: process.env.UNIFORM_PROJECT_ID,
    apiHost: process.env.UNIFORM_CLI_BASE_URL,
    edgeApiHost: process.env.UNIFORM_CLI_BASE_EDGE_URL
  });
  return async function wrappedGetStaticProps(context) {
    var _a2;
    let slugString = resolveSlugFromParams({
      param: options == null ? void 0 : options.param,
      params: context == null ? void 0 : context.params
    });
    if (options == null ? void 0 : options.modifySlug) {
      slugString = options.modifySlug(slugString, context);
    }
    const { preview, previewData } = context;
    let composition = void 0;
    if (previewData == null ? void 0 : previewData.isUniformContextualEditing) {
      composition = { ...import_canvas3.EMPTY_COMPOSITION, _id: (_a2 = previewData.compositionId) != null ? _a2 : import_canvas3.EMPTY_COMPOSITION._id };
    } else {
      try {
        const time = Date.now();
        const response = await canvasClient.getCompositionBySlug({
          slug: slugString,
          state: preview || (options == null ? void 0 : options.preview) ? import_canvas3.CANVAS_DRAFT_STATE : import_canvas3.CANVAS_PUBLISHED_STATE,
          ...options == null ? void 0 : options.requestOptions
        });
        const duration = Date.now() - time;
        composition = response.composition;
        if (!(options == null ? void 0 : options.silent)) {
          (0, import_canvas3.logCompositionResponse)(response, duration);
        }
      } catch (e) {
        console.error("[canvas-next] Failed to fetch composition", e);
        return {
          notFound: true
        };
      }
    }
    const ret = options.callback ? await options.callback(context, composition) : { props: {} };
    if (composition && Object.hasOwn(ret, "props")) {
      const casted = ret;
      casted.props["data"] = composition;
    }
    return ret;
  };
};

// src/slug/index.ts
var getServerSideProps = withUniformGetServerSideProps();
var getStaticProps = withUniformGetStaticProps({ param: "slug" });
var getStaticPaths = withUniformGetStaticPaths();
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getServerSideProps,
  getStaticPaths,
  getStaticProps,
  withUniformGetServerSideProps,
  withUniformGetStaticPaths,
  withUniformGetStaticProps
});
